Devel knot integration

luisfelipemisil opened the pull request:
Create a new knot device from new uplink and join events

A function dedicated to receiving and handling information from new sensors analyzing the current state of the sensors on the knot cloud

luisfelipemisil committed:
Fix the function that creates the knot integration

luisfelipemisil committed:
Create a knot device without a configuration

Create a new knot device when receive a join request or an uplink message from a new lora device OTAA or ABP

luisfelipemisil committed:
Register new devices on the Knot cloud integration

Whenever a new device is created, send a registration request to the knot cloud

If knot cloud returns that the device has already been registered, send an unregistration request to knot cloud

Whenever the knot cloud returns an unregistration confirmation message, check if the device has a token, if it does, it must be deleted. Otherwise,the absence of a token shows that the device was registered in the knot cloud, but did not have its token saved in the chirpstack, so it needs to resubmit a registration request to get a valid token.

luisfelipemisil committed:
Treat the errors arising from the knot cloud

Handle the error of devices that are already registered in the knot cloud but have no token stored in chirp

luisfelipemisil committed:
Fix the error handling function

Remove pointers to strings as json comes with a string

luisfelipemisil committed:
Request Knot Device Authentication

luisfelipemisil committed:
Send the new device configuration data to the knot cloud

Take the configuration data from the Tags in json format and use it to configure the device

Each tag must represent a different sensor, being senseX, tag name, where the 'X' must be the same value as SensorId : X

luisfelipemisil committed:
Standardize the amqp reception code to enter the knot protocol part and fix some details

Correct formatting of comments

Fix proper log type for errors

rtorresmatheus commented on internal/integration/knot/entities/device.go:
```diff
@@ -1,12 +1,27 @@
 package entities
 
+//device state regarding the knot connection
```
This comment is not following the pattern of this file.
```suggestion

const (
	// KNoT Protocol constants
	KnotNew           string = "new"
	KnotRegisterReq          = "registerRequest"
	KnotUnregisterReq        = "registerUnrequest"
	KnotRegistered           = "registered"
	KnotDelete               = "delete"
	KnotForceDelete          = "forceDelete"
	KnotoK                   = "readToSendData"
	KnotAuth                 = "authenticated"
	KnotError                = "error"
)
```

luisfelipemisil replied:
I rewrote this comment in the standard form of the file.

rtorresmatheus commented on internal/integration/knot/knot.go:
```diff
@@ -16,26 +22,65 @@ type Integration struct {
+	DevEUI_str := []byte("")
+	configFrame := entities.Config{}
+	var str string
+	//get all settings from all sensors
```
This comment is not following the pattern of this file.

luisfelipemisil replied:
ok, I removed this comment.

rtorresmatheus commented on internal/integration/knot/knot.go:
```diff
@@ -16,26 +22,65 @@ type Integration struct {
+
+	for _, v := range DevEui {
+		DevEUI_str = strconv.AppendInt(DevEUI_str, int64(v), 16)
+	}
```
This line should be accompanied by a blank line.

luisfelipemisil replied:
ok, I added a new blank line.

rtorresmatheus commented on internal/integration/knot/knot.go:
```diff
@@ -16,26 +22,65 @@ type Integration struct {
+	device := entities.Device{}
+	DevEUI_str := []byte("")
+	configFrame := entities.Config{}
+	var str string
```
This line should be accompanied by a blank line.

luisfelipemisil replied:
ok, I added a new blank line.

rtorresmatheus commented on internal/integration/knot/knot_protocol.go:
```diff
@@ -27,36 +30,92 @@ type protocol struct {
+func (p *protocol) UpdateDevice(device entities.Device) error {
+
+	if _, d := p.devices[device.ID]; !d {
+		//se não existir, retorna false
```
This comment is not following the pattern of this file.

luisfelipemisil replied:
ok, I removed this comment.

rtorresmatheus commented on internal/integration/knot/knot_protocol.go:
```diff
@@ -27,36 +30,92 @@ type protocol struct {
 func (p *protocol) CreateDevice(device entities.Device) error {
+
 	if _, d := p.devices[device.ID]; d {
+
```
This blank line is really necessary?

luisfelipemisil replied:
No, I removed that line.

rtorresmatheus commented on internal/integration/knot/knot_protocol.go:
```diff
@@ -65,3 +124,156 @@ func (p *protocol) DeleteDevice(id string) error {
+
+		// If the device status is new, request a device registration
+		case entities.KnotNew:
+
```
This blank line is really necessary?

luisfelipemisil replied:
In that case, yes, it makes analysis easier.

rtorresmatheus commented on internal/integration/knot/knot_protocol.go:
```diff
@@ -65,3 +124,156 @@ func (p *protocol) DeleteDevice(id string) error {
+
+		// If the device is already registered, ask for device authentication
+		case entities.KnotRegistered:
+
```
This blank line is really necessary?

luisfelipemisil replied:
In that case, yes, it makes analysis easier.

rtorresmatheus commented on internal/integration/knot/knot_protocol.go:
```diff
@@ -65,3 +124,156 @@ func (p *protocol) DeleteDevice(id string) error {
+
+		// Now the device has a token, make a new request for authentication.
+		case entities.KnotAuth:
+
```
This blank line is really necessary?

luisfelipemisil replied:
In that case, yes, it makes analysis easier.

rtorresmatheus commented on internal/integration/knot/knot_protocol.go:
```diff
@@ -65,3 +124,156 @@ func (p *protocol) DeleteDevice(id string) error {
+
+		// Check if the device has a token, if it does, delete it, if not, resend the registration request
+		case entities.KnotDelete:
+
```
This blank line is really necessary?

luisfelipemisil replied:
Yes, it makes analysis easier.

rtorresmatheus commented on internal/integration/knot/knot_protocol.go:
```diff
@@ -65,3 +124,156 @@ func (p *protocol) DeleteDevice(id string) error {
+
+		// Just delete
+		case entities.KnotForceDelete:
+
```
This blank line is really necessary?

luisfelipemisil replied:
Yes, it makes analysis easier.

rtorresmatheus commented on internal/integration/knot/network/subscriber.go:
```diff
@@ -3,9 +3,9 @@ package network
-	bindingKeyRegistered    = "device.registered"
-	bindingKeyUnregistered  = "device.unregistered"
-	bindingKeyUpdatedConfig = "device.config.updated"
+	BindingKeyRegistered    = "device.registered"
```
This is a exported const. These constants should have comment.

luisfelipemisil replied:
Okay, I added a comment for all the constants that will be exported, but I just made one for all of them as they are part of a similar group.

rtorresmatheus commented on internal/integration/knot/network/subscriber.go:
```diff
@@ -24,17 +24,34 @@ func NewMsgSubscriber(amqp *AMQP) Subscriber {
 
 func (ms *msgSubscriber) SubscribeToKNoTMessages(msgChan chan InMsg) error {
```
This is a exported function. This function should have comment.

luisfelipemisil replied:
Okay, I added a new comment for this function.

rtorresmatheus commented on internal/integration/knot/knot_protocol.go:
```diff
@@ -27,36 +30,92 @@ type protocol struct {
+// Update the knot device information on map
+func (p *protocol) UpdateDevice(device entities.Device) error {
+
+	if _, d := p.devices[device.ID]; !d {
```
What does variable d represent? It should have an easy to understand name.

luisfelipemisil replied:
It represents a boolean returned by the function to test the presence or not of the device in this map.
I rewrote this and a few others in a more readable way.

rtorresmatheus reviewed (changes requested):
Great job!
I just left few comments.

luisfelipemisil committed:
Add a method for sending messages to the knot cloud

When a device has gone through the entire knot flow, take the data received and send it to the knot cloud

janael-pinheiro commented on internal/integration/knot/knot.go:
```diff
@@ -16,26 +22,65 @@ type Integration struct {
+var msgChan = make(chan network.InMsg)
+
 // New creates a new KNoT integration.
 func New(m marshaler.Type, conf config.IntegrationKNoTConfig) (*Integration, error) {
```
"New" what? The name of this function is too generic. It would be better to give a more descriptive name. For example, newKNoTIntegration (based on your comment. )

luisfelipemisil replied:
I could change the name, but chirpstack uses this pattern for creating the integrations, so I think it's better to keep that name.

janael-pinheiro replied:
In this case, I agree that it's best to keep the name as it is.

janael-pinheiro commented on internal/integration/knot/knot.go:
```diff
@@ -16,26 +22,65 @@ type Integration struct {
 // New creates a new KNoT integration.
 func New(m marshaler.Type, conf config.IntegrationKNoTConfig) (*Integration, error) {
 	var err error
 	i := Integration{}
```
I suggest avoiding naming variables, constants, structures, etc., with just one letter. Suggestion, rename the variable "i" to "integration" or a more informative name than "i."

luisfelipemisil replied:
Yes, I did as you recommended and changed the name from 'i' to "integration".

janael-pinheiro commented on internal/integration/knot/knot.go:
```diff
@@ -16,26 +22,65 @@ type Integration struct {
+	device := entities.Device{}
+	DevEUI_str := []byte("")
+	configFrame := entities.Config{}
+	var str string
```
I recommend renaming "str" to something more descriptive. Based on the code, "str" could be replaced with "sensorID" or something similar.

luisfelipemisil replied:
Yes, I did as you recommended and changed the name from str to "sensorID".

janael-pinheiro commented on internal/integration/knot/knot.go:
```diff
@@ -16,26 +22,65 @@ type Integration struct {
+	configFrame := entities.Config{}
+	var str string
+	//get all settings from all sensors
+	for i := 1; i < 100; i++ {
```
What does this number 100 mean? Avoid "magic numbers" like 100 in the for loop. It would be preferable to assign that 100 to a constant with a name that describes its meaning. For example, const maxSomething = 100.

luisfelipemisil replied:
This number represents the maximum amount of sensor settings for a knot device in the chirpstack.
I ended up changing this value to a variable with a more suggestive name.

janael-pinheiro commented on internal/integration/knot/knot.go:
```diff
@@ -16,26 +22,65 @@ type Integration struct {
+}
+
 // HandleUplinkEvent sends an UplinkEvent.
 func (i *Integration) HandleUplinkEvent(ctx context.Context, _ models.Integration, vars map[string]string, pl pb.UplinkEvent) error {
```
What is the purpose of the parameters ctx, _, and vars? Are these parameters being used in this function?

luisfelipemisil replied:
ctx is a variable used by chirpstack in these reception handling functions, I believe it is to portray the internal processing up to that point.
_ is to ignore the received value, I need to keep to follow the pattern.
Vars is used to get the variables that chirpstack allows you to add to the device configuration.

janael-pinheiro commented on internal/integration/knot/knot.go:
```diff
@@ -16,26 +22,65 @@ type Integration struct {
+
 	return nil
 }
 
```
Are the HandleAckEvent, HandleErrorEvent. HandleErrorEvent, HandleStatusEvent, HandleStatusEvent, and HandleLocationEvent functions necessary? Have they been completed, or are they in development?

luisfelipemisil replied:
They will be used in future developments.

janael-pinheiro replied:
OK.

janael-pinheiro commented on internal/integration/knot/knot_protocol.go:
```diff
@@ -27,36 +30,92 @@ type protocol struct {
+}
+
 // Close closes the protocol.
 func (p *protocol) Close() error {
```
The name "Close" is too generic. It could look something like "CloseProtocol."

luisfelipemisil replied:
I need to keep as chirpstack default for integrations

janael-pinheiro replied:
I understood.

janael-pinheiro commented on internal/integration/knot/knot_protocol.go:
```diff
@@ -65,3 +124,156 @@ func (p *protocol) DeleteDevice(id string) error {
+	}
+}
+
+// Just formart ther Error msg
```
ther -> the. msg -> message.

luisfelipemisil replied:
ok

janael-pinheiro commented on internal/integration/knot/network/subscriber.go:
```diff
@@ -24,17 +24,34 @@ func NewMsgSubscriber(amqp *AMQP) Subscriber {
+	}
+
+	err = subscribe(msgChan, queueName, exchangeDevice, exchangeTypeDirect, BindingKeyRegistered)
+	if err != nil {
```
This condition is repeated multiple times. Consider putting this condition in a function that takes the variable "err" as a parameter and checks the value of "err".

luisfelipemisil replied:
I need to return something in case there is an error, even using another function to analyze the error, I would need to check the return of this verification function to return the error or not, I couldn't think of another way to do this.

janael-pinheiro replied:
OK.

janael-pinheiro commented on internal/integration/knot/network/amqp.go:
```diff
@@ -49,14 +48,14 @@ func NewAMQP(url string) *AMQP {
+	return nil
 }
 
 // Stop closes the connection started
```
Please remove the commented code.

luisfelipemisil replied:
Ok.

janael-pinheiro commented on internal/integration/knot/knot_protocol.go:
```diff
@@ -65,3 +124,156 @@ func (p *protocol) DeleteDevice(id string) error {
+				log.WithFields(log.Fields{"amqp": "knot"}).Info("received a authentication response")
+			}
+		case network.BindingKeyUpdatedConfig:
+			deviceInf := entities.Device{}
```
deviceInf == "device information"?

luisfelipemisil replied:
Yes.

janael-pinheiro commented on internal/integration/knot/entities/device.go:
```diff
@@ -1,12 +1,27 @@
 package entities
 
+//device state regarding the knot connection
```
"KnotoK" does not follow the naming pattern. It seems to me that the correct one would be "KnotOk".

luisfelipemisil replied:
Ok.

janael-pinheiro commented on internal/integration/knot/knot.go:
```diff
@@ -16,26 +22,65 @@ type Integration struct {
+			break
+		}
+	}
+	fmt.Println(device.Config)
```
Is this snippet still necessary for production code (fmt.Println())?

luisfelipemisil replied:
It is not.

janael-pinheiro reviewed (commented):
Code is fine. I just added a few comments with suggestions for improvements.

luisfelipemisil committed:
Make revision corrections
